### PR TITLE
g.extension: fix installation from gitlab repositories

### PR DESCRIPTION
**Pull request by ninsbl**
Address #623

### CODE REVIEW

**Comment by petrasovaa** [comment]
I didn't look at the PR yet, but I wanted to point out, python requests is not currently grass dependency and it would be great to keep it that way. There was a discussion on that topic some time ago, but I couldn't find it. Is there a way to replace that with standard Python libraries in this case? At least, g.extension should be able to work without it, possibly with some features disabled.

**Comment by ninsbl** [comment]
> I didn't look at the PR yet, but I wanted to point out, python requests is not currently grass dependency and it would be great to keep it that way. There was a discussion on that topic some time ago, but I couldn't find it. Is there a way to replace that with standard Python libraries in this case? At least, g.extension should be able to work without it, possibly with some features disabled.

You are right of course. I had requests installed beforehand and therefore did not notice that this would add a dependency… `requests` is now replaced by the `Request` module from `urllib.request` following the approach here:
https://stackoverflow.com/questions/47594331/python-3-urlopen-http-error-403-forbidden

Adding a user agent solves the 403 error when downloading from gitlab.

File: scripts/g.extension/g.extension.py [hunk 1]
```diff
@@ -1918,7 +1918,7 @@ def resolve_xmlurl_prefix(url, source=None):
         'ignored_suffixes': ['.zip', '.tar.gz', '.tar.bz2', '.tar'],
         'possible_starts': ['', 'https://', 'http://'],
         'url_start': 'https://',
-        'url_end': '/repository/archive.zip',
+        'url_end': '/-/archive/master/{}-master.zip',
```
Comment (wenzeslaus) [inline scripts/g.extension/g.extension.py:1921]: I think something like `...master/{name}-master...` would be better considering that the string is far from the `.format()` call.

File: scripts/g.extension/g.extension.py [hunk 2]
```diff
@@ -1347,11 +1347,11 @@ def download_source_code(source, url, name, outdev,
         # we expect that the module.zip file is not by chance in the archive
         zip_name = os.path.join(tmpdir, 'extension.zip')
         try:
-            response = urlopen(url)
+            response = urlopen(Request(url, headers={'User-Agent': 'Mozilla/5.0'}))
         except URLError:
             grass.fatal(_("Extension <%s> not found") % name)
         with open(zip_name, 'wb') as out_file:
-            shutil.copyfileobj(response, out_file)
+            out_file.write(response.read())
```
Comment (wenzeslaus) [inline scripts/g.extension/g.extension.py:1354]: Just curious, was there are problem with the previous code or is this just for more readability?

File: scripts/g.extension/g.extension.py [hunk 3]
```diff
@@ -2027,9 +2027,9 @@ def resolve_source_code(url=None, name=None):
     GitLab:
 
     >>> resolve_source_code('gitlab.com/JoeUser/GrassModule') # doctest: +SKIP
-    ('remote_zip', 'https://gitlab.com/JoeUser/GrassModule/repository/archive.zip')
+    ('remote_zip', 'https://gitlab.com/JoeUser/GrassModule/'/-/archive/master/GrassModule-master.zip')
     >>> resolve_source_code('https://gitlab.com/JoeUser/GrassModule') # doctest: +SKIP
-    ('remote_zip', 'https://gitlab.com/JoeUser/GrassModule/repository/archive.zip')
+    ('remote_zip', 'https://gitlab.com/JoeUser/GrassModule/'/-/archive/master/GrassModule-master.zip')
```
Comment (wenzeslaus) [inline scripts/g.extension/g.extension.py:2032]: This seems to have a syntax error. Does the doctest run? I'm afraid it does not run as part of the testsuite, but it should be possible to execute using `g.extension --doctest` (a local hack).

**Comment by neteler** [comment]
@ninsbl shall this still go into the upcoming G 7.8.4?

**Comment by ninsbl** [comment]
> @ninsbl shall this still go into the upcoming G 7.8.4?

This PR is superseded by #683, so I am closing this one. Hope that #683 can be merged before 7.8.4, though...

**Pull request closed without merging**